### ISSUE TITLE
Support for schema bumping and migrating resources

### ISSUE THREAD

**Issue by emjburns** [spinnaker/keel#588]
Right now each handler supports only one api version. We should create a way for handlers to support increasing their api version and migrating objects. Kubernetes CRDs [do this](https://kubernetes.io/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definition-versioning/).

My initial thoughts:
- handlers support several versions
- when a resource with an old version is submitted, we convert it to the new version and persist that
- we return that new object back through the api
- we can probably use the git credentials we have to open a PR with that request, or at least surface the migrated schema

Basically, I want this to be really easy to migrate and keep people up to date. This might depend on re-thinking how we interact with git and our current hacky orca stage.

**Comment by emjburns** [comment]
There are several things here:
1. handle increasing a resource api version
2. handle increasing a delivery config api version
3. updating our customers that have an old version stored in code.

The first thing that I'll do is handle increasing a resource api version. This will be only on the api level. If you submit a resource with an old version, we will update it to the new version and save that version. We will also need to have a migration that converts all old resource specs to the new version (on read from the database + in a batch?).

Tentative plan for #1:
- there will be a mapping provided for each resource type and api of the order of versions (for example: for a resource of kind `cluster` and api scope `titus.spinnaker.netflix.com`, first comes v1, then v1.1, then v2)
- there will be a function that migrates a resource from an old version to a new version. I think this function will operate on maps so that we don't need to keep old types around, but not sure. I think this might also be separated from the resource handler - I don't think that it needs to also know about converting types, since it should just deal with the newest type.
- there might be data classes specific to each version of a resource for using in `SubmittedResource` - i.e. there might be lots of types available at the api layer, but we actually only use the latest one internally.
- when a resource is submitted via the api, the resource will be converted to the newest type, and saved that way. the fully converted version will be returned from the api call.
- the same functions will be used when we read resources from the database so that we can convert them to the new version, then re-save them.

**Comment by robfletcher** [comment]
I was slightly inspired in my thinking on this by the way indexeddb works in JavaScript. Basically when you open a connection to the database you pass a version number. When you make changes to your app that require the "schema" to change you'd increment that version number in your code. You then get an `upgradeneeded` event when opening the database connection and can take steps to upgrade the data. It's pretty similar to a Liquibase migration.

The tricky thing is whether we need to keep data classes around for the old versions in order that `SubmittedResource` can still parse them. I don't think that's the end of the world as we would _not_ need to keep handlers around for the old versions, only the models.
